### ISSUE TITLE
Compilation fails on Mac M1

### ISSUE THREAD

**Issue by berikv** [NagyD/SDLPoP#260]
Error output is
```
src $ make
...
gcc  main.o data.o seg000.o seg001.o seg002.o seg003.o seg004.o seg005.o seg006.o seg007.o seg008.o seg009.o seqtbl.o replay.o options.o lighting.o screenshot.o menu.o midi.o opl3.o stb_vorbis.o -o ../prince -L/opt/homebrew/lib -lSDL2_image -lSDL2 -lm
ld: warning: pointer not aligned at address 0x10005C00A (_dialog_settings + 8 from data.o)
ld: warning: pointer not aligned at address 0x10005C002 (_dialog_settings + 0 from data.o)
ld: unaligned pointer(s) for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [../prince] Error 1
```

**Comment by berikv** [comment]
As far as I've researched now, the linker does not like the function pointers in this definition:
```c
// data:2944
extern dialog_settings_type dialog_settings
	INIT(= {
		add_dialog_rect,
		dialog_method_2_frame,
		4, 4, 4, 4, 3, 4, 1
	});
```

Which uses this struct:
```c
struct dialog_type; // (declaration only)
typedef struct dialog_settings_type {
	void (* method_1) (struct dialog_type *dialog);
	void (* method_2_frame) (struct dialog_type *dialog);
	short top_border;
	short left_border;
	short bottom_border;
	short right_border;
	short shadow_bottom;
	short shadow_right;
	short outer_border;
} dialog_settings_type;
```

This is a similar error (but maybe completely unrelated, I'm not super solid on C): https://github.com/curl/curl/issues/7657
The solution looks to be easy in that particular case, however their issue is with a string literal pointer instead of a function pointer: https://github.com/curl/curl/commit/20e980f85b0ea67b0821a807e73d57b281462564

**Comment by berikv** [comment]
It aint pretty, but I got it to work using:

```c
diff --git a/src/data.c b/src/data.c
index 069da98..a6131d8 100644
--- a/src/data.c
+++ b/src/data.c
@@ -24,3 +24,8 @@ The authors of this program may be contacted at https://forum.princed.org
 #undef DATA_H
 #include "data.h"

+dialog_settings_type dialog_settings = {
+    0,
+    0,
+    4, 4, 4, 4, 3, 4, 1
+};
diff --git a/src/data.h b/src/data.h
index 94d005f..f95bb5f 100644
--- a/src/data.h
+++ b/src/data.h
@@ -83,12 +83,7 @@ extern word cplevel_entr[14];
 // data:46C6
 extern dialog_type* copyprot_dialog;
 // data:2944
-extern dialog_settings_type dialog_settings
-       INIT(= {
-               add_dialog_rect,
-               dialog_method_2_frame,
-               4, 4, 4, 4, 3, 4, 1
-       });
+extern dialog_settings_type dialog_settings;
 // data:2B76
 extern rect_type dialog_rect_1 INIT(= {60, 56, 124, 264});
 // data:2B7E
diff --git a/src/main.c b/src/main.c
index f59ba27..cd16e2f 100644
--- a/src/main.c
+++ b/src/main.c
@@ -24,6 +24,9 @@ int main(int argc, char *argv[])
 {
        g_argc = argc;
        g_argv = argv;
+    memcpy(&dialog_settings, add_dialog_rect, sizeof(int32_t));
+    memcpy((void*)(&dialog_settings) + 4, dialog_method_2_frame, sizeof(int32_t));
+
        pop_main();
        return 0;
 }
```

**Comment by berikv** [comment]
I also tested replacing
```c
extern dialog_settings_type dialog_settings
	INIT(= {
		add_dialog_rect,
		dialog_method_2_frame,
		4, 4, 4, 4, 3, 4, 1
	});
```
with just
```c
extern dialog_settings_type dialog_settings INIT(= {0, 0, 4, 4, 4, 4, 3, 4, 1});
```

And that works too: The game plays fine and the dialogs open.
At this point I'm not sure how to test if dialog_settings still works as it should.

**Comment by berikv** [comment]
The dialog code can be tested by going to Settings > Gameplay > Enable copy protection level > (arrow right to turn ON). Then exit the menu and press Shift+L to enter the copy protection level, which shows a dialog.
Whenever the dialog_settings are not set up right, the code will crash there.

I found a simple fix for the misaligned pointer linker error:
```c
extern dialog_settings_type dialog_settings
	INIT(= {
		(void(*)(dialog_type *))(add_dialog_rect),
		(void(*)(dialog_type *))(dialog_method_2_frame),
		4, 4, 4, 4, 3, 4, 1
	});

```

**Comment by berikv** [comment]
Turns out the casting still fails compilation on M1. Not sure how I didn't notice before.

**Comment by berikv** [comment]
Also, memcpy can't work due to _function pointers_ in c not being _objects_
-> https://stackoverflow.com/a/38446372/439096

**Comment by berikv** [comment]
> Also, memcpy can't work due to _function pointers_ in c not being _objects_
> -> https://stackoverflow.com/a/38446372/439096

This turns out to be not exactly correct; Functions are not objects, but function pointers are. Thats why the CR makes function pointers out of the functions and then memcpy's them over.

**Comment by berikv** [comment]
Two great resources on the misaligned pointer issue:
- https://blog.quarkslab.com/unaligned-accesses-in-cc-what-why-and-solutions-to-do-it-properly.html
- https://developer.apple.com/documentation/xcode/misaligned-pointer

**Comment by berikv** [comment]
This issue was resolved in 40e7808
